packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -394,6 +394,52 @@ export default function parse(source: string, state: State): TemplateParseResult
         element.lwc = lwcOpts;
     }
 
+    function applyLwcRenderModeDirective(element: IRElement) {
+        const { parent, tag, attrsList } = element;
+
+        // Can't handle slots in applySlot because it would be too late for class and style attrs
+        if (state.renderMode === LWCDirectiveRenderMode.light && tag === 'slot') {
+            const invalidAttrs = attrsList
+                .filter(({ name }) => name !== 'name')
+                .map(({ name }) => name);
+            if (invalidAttrs.length > 0) {
+                return warnOnElement(
+                    ParserDiagnostics.LWC_LIGHT_SLOT_INVALID_ATTRIBUTES,
+                    element.__original,
+                    [invalidAttrs.join(',')]
+                );
+            }
+        }
+
+        if (parent !== undefined) return;
+
+        const lwcRenderModeAttribute = getTemplateAttribute(
+            element,
+            ROOT_TEMPLATE_DIRECTIVES.RENDER_MODE
+        );
+
+        if (!lwcRenderModeAttribute) {
+            return;
+        }
+
+        removeAttribute(element, ROOT_TEMPLATE_DIRECTIVES.RENDER_MODE);
+
+        if (
+            lwcRenderModeAttribute.type === IRAttributeType.String &&
+            !hasOwnProperty.call(LWCDirectiveRenderMode, lwcRenderModeAttribute.value)
```

Comment:
I wouldn't recommend using `hasOwnProperty` here as it is quite unsafe to match against the key (their name might change) and it confuses Typescript. I would recommend doing it the old fashion and remove the type guard line 440:

```ts
lwcRenderModeAttribute.value !== LWCDirectiveRenderMode.light && lwcRenderModeAttribute.value !== LWCDirectiveRenderMode.shadow
```